source/extensions/filters/http/cache/http_cache.cc

```diff
@@ -74,30 +73,63 @@ LookupRequest::LookupRequest(const Http::RequestHeaderMap& request_headers, Syst
 size_t stableHashKey(const Key& key) { return MessageUtil::hash(key); }
 size_t localHashKey(const Key& key) { return stableHashKey(key); }
 
-// Returns true if response_headers is fresh.
-bool LookupRequest::isFresh(const Http::ResponseHeaderMap& response_headers) const {
-  if (!response_headers.Date()) {
-    return false;
+bool LookupRequest::requiresValidation(const Http::ResponseHeaderMap& response_headers) const {
+  // TODO(yosrym93): Store parsed response cache-control in cache instead of parsing it on every
+  // lookup
+  absl::string_view cache_control =
+      response_headers.getInlineValue(response_cache_control_handle.handle());
+  const ResponseCacheControl response_cache_control(cache_control);
+
+  SystemTime response_time = CacheHeadersUtils::httpTime(response_headers.Date());
+  SystemTime::duration response_age = timestamp_ - response_time;
+
+  ASSERT(response_age >= std::chrono::seconds(0), "Response time is in the future.");
+
+  bool request_max_age_exceeded = request_cache_control_.max_age_.has_value() &&
+                                  request_cache_control_.max_age_.value() < response_age;
+  if (response_cache_control.must_validate_ || request_cache_control_.must_validate_ ||
+      request_max_age_exceeded) {
+    // Either the request or response explicitly require validation or a request max-age requirement
+    // is not satisfied
+    return true;
   }
-  const Http::HeaderEntry* cache_control_header =
-      response_headers.getInline(response_cache_control_handle.handle());
-  if (cache_control_header) {
-    const SystemTime::duration effective_max_age =
-        HttpCacheUtils::effectiveMaxAge(cache_control_header->value().getStringView());
-    return timestamp_ - HttpCacheUtils::httpTime(response_headers.Date()) < effective_max_age;
+
+  // When date metadata injection for responses with no date
+  // is implemented, this ASSERT will need to be updated
+  ASSERT((response_headers.Date() && response_cache_control.max_age_.has_value()) ||
+             response_headers.get(Http::Headers::get().Expires),
+         "Cache entry does not have valid expiration data.");
```

Comment:
again we are having error-checking with an ASSERT. We need to have a story for this behavior in production.

Reply:
This is not error checking, CacheabilityUtils::isCacheableResponse makes sure no responses get cached unless they satisfy this assert.

Reply:
@htuch @asraa for policy clarification on this pattern of asserts.

Reply:
I agree with @yosrym93 here but it would be helpful to add comments to clarify who is providing the guarantee above each `ASSERT`.